### ISSUE TITLE
JetBrains: think about determining file type not based on file extension

### ISSUE THREAD

**Issue by vdavid** [sourcegraph/sourcegraph#55006]
The logic that determines the file type based on the extenion is [here](https://github.com/sourcegraph/sourcegraph/commit/4d9b97d8aae08f0e3c085f9f4aa93930611a85c9#diff-86c0eb7456a931f00456006737ba815ee9d6fca66a0813c01a151bcf5fe30490).

Idea for improvement from @Gedochao:

> :intellij:  doesn't have a strict language id scheme and instead has a variable number of programming languages supported, depending on the installed plugins and whether it's running community, ultimate, webstorm, etc.
> so if you open a JavaScript file in an editor in IntelliJ Community , IntelliJ will identify this as unknown (and even throw a nice null value at me, because why wouldn't it :great_design:)
> Now, I know in recipes we actually identify the language based on the file extension ourselves.
> This is sort of paranoid on my side, but because of the structure mentioned above, it is basically possible to force :intellij: to identify anything as anything - if you have the right plugin, you can make :intellij:  think a .java file has, I dunno HOCON configurations based on context, if you want to for whatever reason.
> So what I mean is, depending on someone's environment, identifying the language based on file extension isn't very accurate, it might not match how the IDE treats the file.
> And then, relying solely on what the JetBrains SDK is returning doesn't necessarily cover all the files we could.
> Merging both approaches isn't perfect, either, because the same language may have different languageIds  depending on how the support is provided in the user's environment.
> TL;DR There are no good, universal solutions, to my knowledge.
> This is an implementation detail, but could lead to errors that might be very difficult to debug in the future (as they may only be replicable when using the exact same environment). Nonetheless, we need to follow some rule here, and whatever gets decided should also reflect on the recipes implementation later.
> What I'm currently doing for the unstable-codegen implementation is trying to use whatever I get from IntelliJ and then fallback to our best guess based on the extension. This seems to be the most reasonable, but could still lead to a mismatch. So that's something to keep in mind.

@vdavid's answer:

> I understand the nuance, but my experience is that in practice, the language is identifiable based on the extension.
> No real good exceptions came to mind, so I asked GPT-4: “Do you have some examples for file types that typically come up in IDEs/code editors where the format of the file is _not_ identifiable from the file type?”
> Its answer was also not that convincing that this is an issue in practice:
> .conf, .cfg, .ini, .properties: These are all configuration files, and they can come in different formats. Some use a standard INI format, some use JSON, some use XML, and some have their own custom format.
> .log: Log files often contain plain text, but the structure and format can vary greatly depending on the application generating the logs.
> .dat: This is a generic data file extension. The format of the data contained within is not identifiable from the file extension.
> .bin, .dump: These are binary files, the contents of which are not human-readable and vary depending on the program that created them.
> .tmp, .temp: These are temporary files. They can contain any kind of data, and thus the format cannot be identified from the file extension.
> Based on this, I’d actually use the file type as the primary source. Then I’d consider edge cases like log files that might often be JSONs in some system, and I’d just mark these file types as “unknown” in our mapping and default to whatever the IDE has for them. And if it doesn’t have anything, just fall back to plain text.
> Bottom line: thanks a lot for thinking about this, I think it could be a very important rabbit hole, but I think we’re lucky this time and can just go with a simple mapping.

@Gedochao:
> Actually, there's a whole thing for files without extensions, where the IDE can recognize it based on the shebang line, for example
> `.conf` files tend to contain wildly different stuff, as configuration syntaxes are a whole other can of worms

→ So we created this issue as a someday-maybe.
